File: typescript/test/Supermarket.test.ts
import {FakeCatalog} from "./FakeCatalog"
import {Product} from "../src/model/Product"
import {SupermarketCatalog} from "../src/model/SupermarketCatalog"
import {Receipt} from "../src/model/Receipt"
import {ShoppingCart} from "../src/model/ShoppingCart"
import {Teller} from "../src/model/Teller"
import {SpecialOfferType} from "../src/model/SpecialOfferType"
import {ProductUnit} from "../src/model/ProductUnit"
import {expect} from "chai";


describe('Supermarket', function () {

    it("two normal items", () => {
        const catalog: SupermarketCatalog = new FakeCatalog();
        const toothbrush: Product = new Product("toothbrush", ProductUnit.Each);
        catalog.addProduct(toothbrush, 0.99);
        const rice: Product = new Product("rice", ProductUnit.Each);
        catalog.addProduct(rice, 2.99);
        const teller: Teller = new Teller(catalog);
        const cart: ShoppingCart = new ShoppingCart();

        // ACT
        cart.addItem(toothbrush);
        cart.addItem(rice)
        const receipt: Receipt = teller.checksOutArticlesFrom(cart);

        // ASSERT
        expect(receipt.getTotalPrice()).closeTo(3.98, 0.01);

    })


    it("two get one free", () => {
        const catalog: SupermarketCatalog = new FakeCatalog();
        const toothbrush: Product = new Product("toothbrush", ProductUnit.Each);
        catalog.addProduct(toothbrush, 0.99);
        const rice: Product = new Product("rice", ProductUnit.Each);
        catalog.addProduct(rice, 2.99);
        const teller: Teller = new Teller(catalog);
        const cart: ShoppingCart = new ShoppingCart();

        // ACT
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, catalog.getUnitPrice(toothbrush));
        const receipt: Receipt = teller.checksOutArticlesFrom(cart);

        // ASSERT
        expect(receipt.getTotalPrice()).closeTo(1.98, 0.01);

    })


    it("x for y discount", () => {
        const catalog: SupermarketCatalog = new FakeCatalog();
        const cherryTomatoes: Product = new Product("cherry Tomato box", ProductUnit.Each);
        catalog.addProduct(cherryTomatoes, 0.69);
        const teller: Teller = new Teller(catalog);
        const cart: ShoppingCart = new ShoppingCart();

        // ACT
        cart.addItem(cherryTomatoes);
        cart.addItem(cherryTomatoes);
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, cherryTomatoes, 0.99);
        const receipt: Receipt = teller.checksOutArticlesFrom(cart);

        // ASSERT
        expect(receipt.getTotalPrice()).closeTo(0.99, 0.01);

    })

    it("five for y discount", () => {
        const catalog: SupermarketCatalog = new FakeCatalog();
        const apples: Product = new Product("apples", ProductUnit.Kilo);
        catalog.addProduct(apples, 1.99);
        const teller: Teller = new Teller(catalog);
        const cart: ShoppingCart = new ShoppingCart();

        // ACT
        cart.addItemQuantity(apples, 5);
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99);
        const receipt: Receipt = teller.checksOutArticlesFrom(cart);

        // ASSERT
        expect(receipt.getTotalPrice()).closeTo(6.99, 0.01);
    })
});

File: kotlin/src/test/kotlin/SupermarketTest.kt

import dojo.supermarket.model.*
import kotlin.test.Test
import kotlin.test.assertEquals


class SupermarketTest {

    @Test
    fun TwoNormalItems_TotalPriceIsSumOfIndividualPrices() {
        val catalog: SupermarketCatalog = FakeCatalog()
        val toothbrush = Product("toothbrush", ProductUnit.EACH)
        catalog.addProduct(toothbrush, 0.99)
        val rice = Product("rice", ProductUnit.EACH)
        catalog.addProduct(rice, 2.99)
        val teller = Teller(catalog)
        val cart = ShoppingCart()

        cart.addItem(toothbrush)
        cart.addItem(rice)
        val receipt = teller.checksOutArticlesFrom(cart)

        assertEquals(3.98, receipt.getTotalPrice(), 0.01)
    }


    @Test
    fun ThreeForTwoOffer_ThirdItemIsFree() {
        val catalog: SupermarketCatalog = FakeCatalog()
        val toothbrush = Product("toothbrush", ProductUnit.EACH)
        catalog.addProduct(toothbrush, 0.99)
        val teller = Teller(catalog)
        val cart = ShoppingCart()

        cart.addItem(toothbrush)
        cart.addItem(toothbrush)
        cart.addItem(toothbrush)
        teller.addSpecialOffer(SpecialOfferType.THREE_FOR_TWO, toothbrush, catalog.getUnitPrice(toothbrush))
        val receipt = teller.checksOutArticlesFrom(cart)

        assertEquals(1.98, receipt.getTotalPrice(), 0.01)
    }

    @Test
    fun TwoForAmountOffer_PriceIsReducedToBundlePrice() {
        val catalog: SupermarketCatalog = FakeCatalog()
        val cherryTomatoes = Product("cherry Tomato box", ProductUnit.EACH)
        catalog.addProduct(cherryTomatoes, 0.69)
        val teller = Teller(catalog)
        val cart = ShoppingCart()

        cart.addItem(cherryTomatoes)
        cart.addItem(cherryTomatoes)
        teller.addSpecialOffer(SpecialOfferType.TWO_FOR_AMOUNT, cherryTomatoes, 0.99)
        val receipt = teller.checksOutArticlesFrom(cart)

        assertEquals(0.99, receipt.getTotalPrice(), 0.01)
    }

    @Test
    fun FiveForAmountOffer_PriceIsReducedToBundlePrice() {
        val catalog: SupermarketCatalog = FakeCatalog()
        val apples = Product("apples", ProductUnit.KILO)
        catalog.addProduct(apples, 1.99)
        val teller = Teller(catalog)
        val cart = ShoppingCart()

        cart.addItemQuantity(apples, 5.0)
        teller.addSpecialOffer(SpecialOfferType.FIVE_FOR_AMOUNT, apples, 6.99)
        val receipt = teller.checksOutArticlesFrom(cart)

        assertEquals(6.99, receipt.getTotalPrice(), 0.01)
    }
}